packages/router/src/provide_router.ts

```
@@ -203,7 +203,13 @@ export function getBootstrapListener() {
     const bootstrapDone = injector.get(BOOTSTRAP_DONE);
 
     if (injector.get(INITIAL_NAVIGATION) === InitialNavigation.EnabledNonBlocking) {
-      router.initialNavigation();
+      const errorHandler = injector.get(ErrorHandler, null);
+      const initialNavigation = router.initialNavigation();
+      if (errorHandler) {
```

The error would already be surfaced through the router’s errorHandler as well as the events as a NavigationError, right? Do we really need it raised in a third place?

---

No the error is not captured in any other place. 

At the moment it is surfaced as an unhandled promise rejections, this just captures it and pass it to the error handler.

---

Actually I stand corrected, the error is being added as navigation error. So the problem here is that the error handler provided from the DI is not used.

What are your thoughts about using the error handler provided through the DI token (ERROR_HANDLER) instead of an option to the router?  Something like;

```ts
  private coreErrorHandler = inject(ERROR_HANDLER);

  /**
   * A handler for navigation errors in this NgModule.
   *
   * @deprecated Subscribe to the `Router` events and watch for `NavigationError` instead.
   *   `provideRouter` has the `withNavigationErrorHandler` feature to make this easier.
   * @see `withNavigationErrorHandler`
   */
  errorHandler = this.options.errorHandler ?? function defaultErrorHandler(error: unknown): never {
     this.coreErrorHandler.handle(error);
     throw error;
  }
```

This is important so not to make platform-server dependent on the router and have to listen to navigation errors from the router directly, where when using SSG we do want navigation errors to be fatal and causes the process to exit with a non zero error code.

---

So I would actually be more in favor of the initialNavigation catching and swallowing the promise rejection since it’s surfaced elsewhere. That should resolve the issue as well, right?

---

The problem with catching the error will be swallowed since errors captured as `NavigationErrors` by default are not surfaced to the users. This is problematic as “critical” errors such as during component construction will not be surfaced unless users subscribe to the router events. 

I see a couple of possible solutions. 
1. Catch the rejection and log it. 
```ts
this.scheduleNavigation(urlTree, source, restoredState, extras).catch(e => this.errorHandler.handleError(e));
```

2. Catch the rejection and change the `defaultErrorHandler` and capture log it  
```ts
private coreErrorHandler = inject(ERROR_HANDLER);

errorHandler = this.options.errorHandler ?? function defaultErrorHandler(error: unknown): never {
   this.coreErrorHandler.handle(error);
   throw error;
}
...
this.scheduleNavigation(urlTree, source, restoredState, extras).catch(() => {});
```

The above somewhat aligns with what you want to do in https://github.com/angular/angular/pull/48910 but it does not remove the `throw` yet.

3. Catch the error and register a default `NavigationErrorHandler` when none is provided. 
```ts
export function provideRouter(routes: Routes, ...features: RouterFeatures[]): EnvironmentProviders {
  const providers: Provider[] = [
    {provide: ROUTES, multi: true, useValue: routes},
    typeof ngDevMode === 'undefined' || ngDevMode
      ? {provide: ROUTER_IS_PROVIDED, useValue: true}
      : [],
    {provide: ActivatedRoute, useFactory: rootRoute, deps: [Router]},
    {provide: APP_BOOTSTRAP_LISTENER, multi: true, useFactory: getBootstrapListener},
    features.map((feature) => feature.ɵproviders),
  ];

  if (!features.some(f => f.ɵkind === RouterFeatureKind.NavigationErrorHandlerFeature)) {
    providers.push(
      withNavigationErrorHandler((e) => inject(ErrorHandler).handleError(e)).ɵproviders
    );
  }

  return makeEnvironmentProviders(providers);
}
....
this.scheduleNavigation(urlTree, source, restoredState, extras).catch(() => {});
```